### PR TITLE
[FLINK-17306] Add open method to (De)SerializationSchema

### PR DESCRIPTION
**Pull request by dawidwys**
## What is the purpose of the change

This PR adds open method to the `DeserializationSchema` and `SerializationSchema`. It gives users access to metrics. Moreover it adds a well defined step in the lifecycle to perform initialization.

It also updates Kafka, Kinesis, PubSub and GCP connectors to call the method.

## Verifying this change

Check tests in corresponding connectors.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

### CODE REVIEW

**Comment by dawidwys** [comment]
cc @aljoscha @kl0u

File: flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/KeyedSerializationSchemaWrapper.java [hunk 1]
```diff
@@ -37,6 +37,10 @@ public KeyedSerializationSchemaWrapper(SerializationSchema<T> serializationSchem
 		this.serializationSchema = serializationSchema;
 	}
 
+	public SerializationSchema<T> getSerializationSchema() {
```
Comment (aljoscha) [inline flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/KeyedSerializationSchemaWrapper.java:40]: Why don't we forward the `open()` call but need to do casts when invoking?

Comment (dawidwys) [reply flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/KeyedSerializationSchemaWrapper.java:40]: I did not want to add new features to a deprecated interface. The KeyedSerializationSchema is deprecated and that's why I did not want to add the `open()` method there. Therefore there is no method where I can forward it. Still I wanted to support the open for the top-level `SerializationSchema`, which is wrapped in that deprecated class.

Comment (aljoscha) [reply flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/KeyedSerializationSchemaWrapper.java:40]: Ok, that seems fine. Good to merge then.

**Comment by dawidwys** [comment]
Thank you for the review.

**Pull request merged**